### ISSUE TITLE
[Canary] Stable(v1.0.0) and Canary(v1.1.0) performance differences

### ISSUE THREAD

**Issue by Jolomann** [Quaver/Quaver#3674]
I recently had performance issues with canary so I ran some benchmarks to see exactly what the difference was. The other day I had some awful frame skips I can't reproduce so I'm assuming the skips weren't game related

For benchmarking I used the default bar skin and autoplayed a ranked 7k map with many objects on 2.0x (https://quavergame.com/mapset/map/7753). Since stutters weren't reproducible I only did a single pass on both clients to show fps differences

Benchmarks performed with CapFrameX, starting capture by pressing enter and the capture key simultaneously when entering gameplay
![](https://cdn.discordapp.com/attachments/461621899320098816/1032574674514739280/unknown.png)

Filtered data*
![Quaver_ComparisonChartExport_fps](https://user-images.githubusercontent.com/33073129/196900953-ed88cd23-0de3-4e2a-be72-bebc932d10b7.svg)

Raw data*
![RAWQuaver_ComparisonChartExport_fps](https://user-images.githubusercontent.com/33073129/196902449-c27721cc-bd50-4803-8bc0-559ad84d6372.svg)
*Start and end peaks are out of frame because screen transition data isn't relevant here

Raw data reveals that stable has more variance, and looking at sensor data shows why. It seems there's more GPU overhead during the canary run, indicating a CPU bottleneck. GPU load, power draw, frequency, and temps were lower on canary

Stable sensor data
![image](https://user-images.githubusercontent.com/33073129/196903717-0e1de5b7-7593-4775-821a-b943bae76de9.png)

Canary sensor data
![image](https://user-images.githubusercontent.com/33073129/196903681-c98872e9-364b-4287-85a5-cb9c94e8840b.png)

**Specs**
CPU: AMD Ryzen 5 3600X
GPU: NVIDIA GeForce GTX 970
RAM: 32GB DDR4 2866MHz
Windows 10

**Additional context**
If needed, these files can be imported in CapFrameX for closer inspection
https://cdn.discordapp.com/attachments/461621899320098816/1032574349221298176/CapFrameX-Quaver.exe-2022-10-20T92239.json
https://cdn.discordapp.com/attachments/461621899320098816/1032574349720440893/CapFrameX-Quaver.exe-2022-10-20T92746.json

**Comment by Jolomann** [comment]
It also appears fps decreases over time in a song but that's consistent on both clients so consider that a separate issue

**Comment by Illuminati-CRAZ** [comment]
I'm thinking this is probably related to #3658. I think what's happening is that the hitobject sprites became more expensive to draw for some reason. In the heavy SV scenario I could correlate drops in framerate to an increase of hitobjects in the hitobject pool, and it seems to me that the continuous slight decrease in framerate here could be caused by a continuous slight increase in pool size.

**Comment by Swan** [comment]
@Jolomann  @Illuminati-CRAZ Try the latest update to see if this is still happening. If it is, I'll investigate more.

**Comment by Illuminati-CRAZ** [comment]
Performance is still poor

Watching [Drater's Elevator map](https://quavergame.com/mapset/map/71929) at 26 seconds in song select map preview:
Steam client - ~50 fps
Steam canary - Game becomes unresponsive
Local Public configuration build - ~50 fps
Local Release configuration build - ~150 fps

**Comment by Swan** [comment]
> Watching [Drater's Elevator map](https://quavergame.com/mapset/map/71929) at 26 seconds in song select map preview: 

Something about this specific SV effect is causing it. Need to take a look into what exactly what it's doing. Unless you have more insight.

```
- StartTime: 25914
  Multiplier: 0.75
- StartTime: 26414
  Multiplier: 268435456
- StartTime: 26414.015625
- StartTime: 26415
  Multiplier: -19200
- StartTime: 26415.015625
- StartTime: 26812.984375
  Multiplier: 19200
- StartTime: 26813
```

**Comment by Illuminati-CRAZ** [comment]
The effect causes position to teleport forward about 420 million units, which covers pretty much the whole map. I assume the effect is causing a bunch of hitobjects to be created for the hitobject pool, and the game is doing all the math for each hitobject and its sprite. This would also be consistent with the observation that seeking forwards or backwards in the preview causes the performance to recover, as doing so reset the object pool.

**Comment by Jolomann** [comment]
Upgraded GPU since last time so not a perfect comparison to the last test. There's still a notable performance difference but it's definitely looking better. The GPU upgrade shouldn't be that relevant since it wasn't being utilised on the last test. On this test the GPU stats were within margin of error instead of completely different
![image](https://user-images.githubusercontent.com/33073129/205517736-edeb9d95-1d7a-4e55-84a7-10842b1d00c6.png)

**Comment by robby250** [comment]
chiming in to say that im getting lag spikes every few seconds on the latest canary version, where the previous version was fine, and the stable version is fine too (theres probably some memory leaks on stable too cuz the game starts lagging after a few hours but ... at least im not getting spikes as soon as i start the game)

**Comment by Swan** [comment]
> chiming in to say that im getting lag spikes every few seconds on the latest canary version, where the previous version was fine,

Good to know. That's what I was testing on the latest build.

**Comment by Illuminati-CRAZ** [comment]
Fixed with v1.2.2
